File: ios/exe/Packages/ExeDomain/Sources/ExeDomain/UserProfile.swift
import Foundation

public struct LinkedSlackUser: Codable, Hashable, Sendable {
    public let slackTeamId: SlackTeamID
    public let slackUserId: SlackUserID
    public let workspaceId: WorkspaceID

    public init(
        slackTeamId: SlackTeamID,
        slackUserId: SlackUserID,
        workspaceId: WorkspaceID
    ) {
        self.slackTeamId = slackTeamId
        self.slackUserId = slackUserId
        self.workspaceId = workspaceId
    }
}

public struct UserProfile: Codable, Hashable, Identifiable, Sendable {
    public let createdAt: DateTime
    public let displayName: String?
    public let email: String
    public let id: UserID
    public let slackUsers: [LinkedSlackUser]
    public let updatedAt: DateTime
    public let workspaceIds: [WorkspaceID]

    public init(
        createdAt: DateTime,
        displayName: String?,
        email: String,
        id: UserID,
        slackUsers: [LinkedSlackUser],
        updatedAt: DateTime,
        workspaceIds: [WorkspaceID]
    ) {
        self.createdAt = createdAt
        self.displayName = displayName
        self.email = email
        self.id = id
        self.slackUsers = slackUsers
        self.updatedAt = updatedAt
        self.workspaceIds = workspaceIds
    }
}

File: app/packages/domain/src/user-profile.ts
import {
  dateTimeSchema,
  slackTeamIdSchema,
  slackUserIdSchema,
  userIdSchema,
  workspaceIdSchema,
} from './common';
import { z } from 'zod';

export const linkedSlackUserSchema = z
  .object({
    slackTeamId: slackTeamIdSchema,
    slackUserId: slackUserIdSchema,
    // ISO timestamp of the last live Slack membership confirmation. Optional
    // for backward compatibility with profiles written before live
    // verification existed; treated as "never verified" (stale) when absent.
    verifiedAt: dateTimeSchema.optional(),
    workspaceId: workspaceIdSchema,
  })
  .strict();

export const userProfileSchema = z
  .object({
    createdAt: dateTimeSchema,
    displayName: z.string().min(1).optional(),
    email: z.email(),
    id: userIdSchema,
    // ISO timestamp of the last FULL workspace discovery (a scan of every
    // workspace) at login. Used to bound staleness of the candidate index:
    // once older than the discovery TTL, login re-scans all workspaces instead
    // of trusting the index, catching memberships the index never learned of.
    // Optional for backward compatibility (absent => treated as never).
    lastFullDiscoveryAt: dateTimeSchema.optional(),
    slackUsers: z.array(linkedSlackUserSchema),
    updatedAt: dateTimeSchema,
    workspaceIds: z.array(workspaceIdSchema),
  })
  .strict();

export type LinkedSlackUser = z.infer<typeof linkedSlackUserSchema>;

export type UserProfile = z.infer<typeof userProfileSchema>;
